src/lib/downloader/downloader.ml

```diff
@@ -232,184 +258,276 @@ end = struct
-          not (Hash_set.is_empty to_try) )
+      Hashtbl.iter t.knowledge ~f:(fun s ->
+          List.iter ks ~f:(Hash_set.remove s.tried_and_failed) )
+
+    let ignore_period = Time.Span.of_min 2.
 
     let update t u =
       match u with
+      | Add_knowledge {peer; claimed; out_of_band} ->
+          if not out_of_band then
+            Hash_set.remove t.knowledge_requesting_peers peer ;
+          Hashtbl.update t.knowledge peer ~f:(function
+            | None ->
+                { Knowledge.claimed= Some (`Some claimed)
+                ; tried_and_failed= Key.Hash_set.create () }
+            | Some k ->
+                let claimed =
+                  match k.claimed with
+                  | None ->
+                      `Some claimed
+                  | Some (`Some claimed') ->
+                      `Some
+                        (List.dedup_and_sort ~compare:Key.compare
+                           (claimed' @ claimed))
+                  | Some `All ->
+                      `All
+                  | Some (`Call f) ->
+                      let s = Key.Hash_set.of_list claimed in
+                      `Call (fun key -> f key || Hash_set.mem s key)
+                in
+                {k with claimed= Some claimed} )
+      | Knowledge_request_starting peer ->
+          Hash_set.add t.knowledge_requesting_peers peer
+      | Knowledge {peer; claimed; active_jobs; out_of_band} ->
+          if not out_of_band then
+            Hash_set.remove t.knowledge_requesting_peers peer ;
+          let tried_and_failed =
+            let s = Key.Hash_set.create () in
+            List.iter active_jobs ~f:(fun j ->
+                match Map.find j.J.attempts peer with
+                | None ->
+                    ()
+                | Some a ->
+                    if not (Attempt.worth_retrying a) then Hash_set.add s j.key
+            ) ;
+            s
+          in
+          Hashtbl.set t.knowledge ~key:peer
+            ~data:{Knowledge.claimed= Some claimed; tried_and_failed}
       | Job_cancelled h ->
           jobs_no_longer_needed t [h] ;
-          Hashtbl.iter t.knowledge ~f:(fun s -> Hash_set.remove s h)
-      | Download_starting (peer, ks) ->
-          (* WHen a download starts for a job, we should remove all peers from
-           available and useful whose only useful jobs are in downloading.. *)
-          Hash_set.add t.downloading_peers peer ;
-          List.iter ks ~f:(Hash_set.add t.downloading_keys) ;
-          Available_and_useful.remove t.available_and_useful peer ;
-          Available_and_useful.don't_need_keys t.available_and_useful ks
-      | Download_finished (peer0, `Successful succs, `Unsuccessful unsuccs) ->
-          if not (List.is_empty succs) then Hash_set.add t.all_preferred peer0 ;
+          Hashtbl.iter t.knowledge ~f:(fun s ->
+              Hash_set.remove s.tried_and_failed h )
+      | Download_starting peer ->
+          Hash_set.add t.downloading_peers peer
```

Comment:
all_preferred contains all the preferred peers (even the ones downloading) so that we know when we are in the case of "we could make progress but all the useful peers are currently downloading something"